builtin/providers/ultradns/resource_ultradns_dirpool.go

```diff
@@ -0,0 +1,612 @@
+	res.Codes = make([]string, 0, len(rawCodes))
+	for _, i := range rawCodes {
+		res.Codes = append(res.Codes, i.(string))
+	}
+	return res, err
+}
+
+// makeIPInfo converts a map[string]interface{} from an ip_info block
+// into an IPInfo
+func makeIPInfo(configured interface{}) (udnssdk.IPInfo, error) {
+	var res udnssdk.IPInfo
+	c := configured.(map[string]interface{})
+	err := mapDecode(c, &res)
+	if err != nil {
+		return res, err
+	}
+
+	rawIps := c["ips"].(*schema.Set).List()
+	res.Ips = make([]udnssdk.IPAddrDTO, 0, len(rawIps))
+	for _, rawIa := range rawIps {
+		var i udnssdk.IPAddrDTO
+		err = mapDecode(rawIa, &i)
+		if err != nil {
+			return res, err
+		}
+		res.Ips = append(res.Ips, i)
+	}
+	return res, nil
+}
+
+// collate and zip RData and RDataInfo into []map[string]interface{}
+func zipDirpoolRData(rds []string, rdis []udnssdk.DPRDataInfo) []map[string]interface{} {
+	result := make([]map[string]interface{}, 0, len(rds))
+	for i, rdi := range rdis {
+		r := map[string]interface{}{
+			"host":               rds[i],
+			"all_non_configured": rdi.AllNonConfigured,
+			"ip_info":            mapFromIPInfos(rdi.IPInfo),
+			"geo_info":           mapFromGeoInfos(rdi.GeoInfo),
+		}
+		result = append(result, r)
+	}
+	return result
+}
+
+// makeSetFromDirpoolRdata encodes an array of Rdata into a
+// *schema.Set in the appropriate structure for the schema
+func makeSetFromDirpoolRdata(rds []string, rdis []udnssdk.DPRDataInfo) *schema.Set {
+	s := &schema.Set{F: hashRdatas}
+	rs := zipDirpoolRData(rds, rdis)
+	for _, r := range rs {
+		s.Add(r)
+	}
+	return s
+}
+
+// mapFromIPInfos encodes 0 or 1 IPInfos into a []map[string]interface{}
+// in the appropriate structure for the schema
+func mapFromIPInfos(rdi *udnssdk.IPInfo) []map[string]interface{} {
+	res := make([]map[string]interface{}, 0, 1)
+	if rdi != nil {
+		m := map[string]interface{}{
+			"name":             rdi.Name,
+			"is_account_level": rdi.IsAccountLevel,
+			"ips":              makeSetFromIPAddrDTOs(rdi.Ips),
+		}
+		res = append(res, m)
+	}
+	return res
+}
+
+// makeSetFromIPAddrDTOs encodes an array of IPAddrDTO into a
+// *schema.Set in the appropriate structure for the schema
+func makeSetFromIPAddrDTOs(ias []udnssdk.IPAddrDTO) *schema.Set {
+	s := &schema.Set{F: hashIPInfoIPs}
+	for _, ia := range ias {
+		s.Add(mapEncode(ia))
+	}
+	return s
+}
+
+// mapFromGeoInfos encodes 0 or 1 GeoInfos into a []map[string]interface{}
+// in the appropriate structure for the schema
+func mapFromGeoInfos(gi *udnssdk.GeoInfo) []map[string]interface{} {
+	res := make([]map[string]interface{}, 0, 1)
+	if gi != nil {
+		m := mapEncode(gi)
+		m["codes"] = makeSetFromStrings(gi.Codes)
+		res = append(res, m)
+	}
+	return res
+}
+
+// hashIPInfoIPs generates a hashcode for an ip_info.ips block
+func hashIPInfoIPs(v interface{}) int {
+	var buf bytes.Buffer
+	m := v.(map[string]interface{})
+	buf.WriteString(fmt.Sprintf("%s-", m["start"].(string)))
```

Comment:
if these are not all required then we should check to make sure there is a value before adding to the hash IMO